### PR TITLE
Allow super users to update shared flag on zones

### LINKED ISSUE #348: Allow super users to update a zone to shared (or unshared)
Similar logically to #345 

For now, this functionality is super user only.

See `ZoneInput` type in #345. Make update zone endpoint take in that type.

Logically, need to compare old with new `shared` field. If None is passed in, keep old `shared`. If something is passed in and its different than the old status, fail for normal user, succeed for super user.

---
similarly, this is hard to func test, but can validate that a normal user can not update a non-shared zone to be shared.

### PR DESCRIPTION
**Pull request by mitruly**
Allow super users to update shared flag on zones.
Fixes #348.

Changes in this pull request:
- Create `UpdateZoneInput` model for update request parameters. Update zone routing, validations and service accordingly.
- Remove `Zone` and `ZoneChange` JSON deserializers (since we no longer accept either as valid inputs).
- Updated tests.
- Minor documentation update.

### CODE REVIEW

File: modules/api/functional_test/live_tests/zones/update_zone_test.py [hunk 1]
```diff
@@ -819,3 +821,8 @@ def test_update_zone_no_authorization(shared_zone_test_context):
     }
 
     client.update_zone(zone, sign_request=False, status=401)
+
+def test_normal_user_cannot_update_shared_zone_flag(shared_zone_test_context):
+    """
+    Test updating a zone shared status as a normal user fails
```
Comment (rebstar6) [inline modules/api/functional_test/live_tests/zones/update_zone_test.py]: cool test

File: modules/api/functional_test/live_tests/zones/update_zone_test.py [hunk 2]
```diff
@@ -191,11 +192,11 @@ def test_update_invalid_zone_data(shared_zone_test_context):
             'id': result_zone['id'],
             'name': result_zone['name'],
             'email': 'test@test.com',
-            'status': 'invalid_status'
+            'adminGroupId': True
         }
 
         errors = client.update_zone(update_zone, status=400)['errors']
-        assert_that(errors, contains_inanyorder('Invalid ZoneStatus'))
+        assert_that(errors, contains_inanyorder('Do not know how to convert JBool(true) into class java.lang.String'))
```
Comment (britneywright) [inline modules/api/functional_test/live_tests/zones/update_zone_test.py]: Why was the Invalid ZoneStatus error assertion removed? Both errors should appear right?

File: modules/docs/src/main/tut/api/zone-model.md [hunk 3]
```diff
@@ -30,9 +30,10 @@ account       | string      | **DEPRECATED** The account that created the zone |
 email         | string      | The distribution email for the zone |
 connection    | ZoneConnection | The connection used to issue DDNS updates to the backend zone.  If not provided, default keys will be used.  See the [Zone Connection Attributes](#zone-conn-attr) for more information |
 transferConnection | ZoneConnection | The connection that is used to sync the zone with the DNS backend.  This can be different than the update connection.  If not provided, default keys will be used |
-shared        | boolean     | **DEPRECATED** An indicator that the zone is shared with anyone.  No longer used as ACLs are now used to manage zones. |
+shared        | boolean     | An indicator that the zone is shared with anyone. |
```
Comment (britneywright) [inline modules/docs/src/main/tut/api/zone-model.md]: Should we indicate this has limited use at this time? Like it can only be set by super users.

File: modules/docs/src/main/tut/api/zone-model.md [hunk 4]
```diff
@@ -30,9 +30,10 @@ account       | string      | **DEPRECATED** The account that created the zone |
...
 acl           | ZoneACL     | The access control rules governing the zone.  See the [Zone ACL Rule Attributes](#zone-acl-rule-attr) for more information
 id            | string      | The unique identifier for this zone
+isTest        | boolean     | An indicator that the zone is a zone for testing purposes. |
```
Comment (britneywright) [inline modules/docs/src/main/tut/api/zone-model.md]:
```suggestion
 isTest        | boolean     | Indicates the zone is intended for testing purposes only if true. |
```

File: modules/api/functional_test/live_tests/zones/update_zone_test.py (hunk 2 above)
Comment (mitruly) [reply modules/api/functional_test/live_tests/zones/update_zone_test.py]: `ZoneStatus` is no longer accepted as part of the zone input entities (in either create or update), so we don't bother deserializing it any more.

File: modules/docs/src/main/tut/api/zone-model.md (hunk 3 above)
Comment (mitruly) [reply modules/docs/src/main/tut/api/zone-model.md]: That's a good suggestion - I'll add a tidbit about who can set/modify this.

File: modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneService.scala [hunk 5]
```diff
@@ -59,42 +59,51 @@ class ZoneService(
...
-      updateZoneChange <- ZoneChangeGenerator.forUpdate(newZone, existingZone, auth).toResult
+      _ <- isValidZoneAcl(updateZoneInput.acl).toResult
+      existingZone <- getZoneOrFail(updateZoneInput.id)
+      _ <- validateSharedZoneAuthorized(
+        existingZone.shared,
+        updateZoneInput.shared,
+        auth.signedInUser).toResult
+      _ <- canChangeZone(auth, existingZone.name, existingZone.adminGroupId).toResult
+      _ <- adminGroupExists(updateZoneInput.adminGroupId)
+      _ <- canChangeZone(auth, updateZoneInput.name, updateZoneInput.adminGroupId).toResult // if admin group changes,
```
Comment (rebstar6) [inline modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneService.scala]: comment multiline here got weird, maybe move it all above this line?

File: modules/api/functional_test/live_tests/zones/update_zone_test.py [hunk 6]
```diff
@@ -819,3 +821,30 @@ def test_update_zone_no_authorization(shared_zone_test_context):
...
+        zone = {
+            'name': zone_name,
+            'email': 'test@test.com',
+            'adminGroupId': shared_zone_test_context.ok_group['id']
+        }
+
+        result = client.create_zone(zone, status=202)
+        result_zone = result['zone']
+        client.wait_until_zone_exists(result)
+
```
Comment (rebstar6) [inline modules/api/functional_test/live_tests/zones/update_zone_test.py]: for good measure, you should be able test the super user in your local by making testUser have super=true, and sending create/update zone requests in postman

can you do that and just make sure this all works as expected (if you havent yet)

File: modules/core/src/main/scala/vinyldns/core/domain/zone/Zone.scala [hunk 7]
```diff
@@ -100,6 +115,17 @@ final case class CreateZoneInput(
     adminGroupId: String,
     isTest: Boolean = false)
 
+case class UpdateZoneInput(
```
Comment (rebstar6) [inline modules/core/src/main/scala/vinyldns/core/domain/zone/Zone.scala]: final?

File: modules/api/src/test/scala/vinyldns/api/domain/zone/ZoneServiceSpec.scala [hunk 8]
```diff
@@ -71,6 +71,19 @@ class ZoneServiceSpec
     new ZoneValidations(1000),
     AccessValidations)
 
+  private val createZoneAuthorized = CreateZoneInput(
+    "ok.zone.recordsets.",
+    "test@test.com",
+    connection = testConnection,
+    adminGroupId = grp.id)
```
Comment (rebstar6) [inline modules/api/src/test/scala/vinyldns/api/domain/zone/ZoneServiceSpec.scala]: ugh our test data.

are grp.id and okGroup.id different?

Comment (rebstar6) [reply modules/api/src/test/scala/vinyldns/api/domain/zone/ZoneServiceSpec.scala]: note were moving towards the TestMembershipData (etc) in core. if you import that (._), does just using okGroup and twoUserGroup here work just as well?

if so i say do that

File: modules/api/src/test/scala/vinyldns/api/domain/zone/ZoneServiceSpec.scala [hunk 9]
```diff
@@ -198,21 +216,53 @@ class ZoneServiceSpec
...
+
+    "succeed if zone shared flag is updated and user is a super user" in {
+      val newZone = updateZoneAuthorized.copy(shared = false)
+      doReturn(IO.pure(Some(Zone(createZoneAuthorized.copy(shared = true)))))
+        .when(mockZoneRepo)
+        .getZone(newZone.id)
+
+      val result = rightResultOf(
+        underTest
+          .updateZone(newZone, okAuth.copy(signedInUser = okUser.copy(isSuper = true)))
```
Comment (rebstar6) [inline modules/api/src/test/scala/vinyldns/api/domain/zone/ZoneServiceSpec.scala]: AuthPrincipal(TestMembershipData.superUser, List.empty)

File: modules/api/src/test/scala/vinyldns/api/domain/zone/ZoneServiceSpec.scala [hunk 10]
```diff
@@ -198,21 +216,53 @@ class ZoneServiceSpec
...
+      val newZone = updateZoneAuthorized.copy(shared = false)
+      doReturn(IO.pure(Some(Zone(createZoneAuthorized.copy(shared = true)))))
+        .when(mockZoneRepo)
+        .getZone(newZone.id)
+
+      val error = leftResultOf(underTest.updateZone(newZone, okAuth).value)
+      error shouldBe a[NotAuthorizedError]
+    }
+
+    "succeed if zone is shared and user is not a super user" in {
```
Comment (rebstar6) [inline modules/api/src/test/scala/vinyldns/api/domain/zone/ZoneServiceSpec.scala]: `succeed if zone shared flag is unchanged and user is not a super user`

File: modules/docs/src/main/tut/api/create-zone.md [hunk 11]
```diff
@@ -15,7 +15,7 @@ if no info is provided the default VinylDNS connections will be used
 
 #### HTTP REQUEST PARAMS
 
-**zone fields**  - adminGroupId, name, and email are required - refer to [zone model](../api/zone-model) |
+**zone fields**  - `adminGroupId`, `name`, and `email` are required - refer to [zone model](../api/zone-model). `shared` (default: `false`) and `isTest` (default: `false`) are optional parameters. |
```
Comment (rebstar6) [inline modules/docs/src/main/tut/api/create-zone.md]: leave out isTest i think. or make it clear that they should not be using that, its just for automated testing with pre-loaded users

File: modules/docs/src/main/tut/api/zone-model.md [hunk 12]
```diff
@@ -30,9 +30,10 @@ account       | string      | **DEPRECATED** The account that created the zone |
 email         | string      | The distribution email for the zone |
 connection    | ZoneConnection | The connection used to issue DDNS updates to the backend zone.  If not provided, default keys will be used.  See the [Zone Connection Attributes](#zone-conn-attr) for more information |
 transferConnection | ZoneConnection | The connection that is used to sync the zone with the DNS backend.  This can be different than the update connection.  If not provided, default keys will be used |
-shared        | boolean     | **DEPRECATED** An indicator that the zone is shared with anyone.  No longer used as ACLs are now used to manage zones. |
+shared        | boolean     | An indicator that the zone is shared with anyone. Can only be set and/or modified by super users. Default value is `false`. |
```
Comment (rebstar6) [inline modules/docs/src/main/tut/api/zone-model.md]: actually, can you split these docs changes into a separate PR? i wanna wait on that merge until our support is complete

File: modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneService.scala [hunk 13]
```diff
@@ -59,42 +59,51 @@ class ZoneService(
...
+      // this confirms user has access to new group
+      zoneToUpdate = Zone(updateZoneInput, existingZone)
+      _ <- validateZoneConnectionIfChanged(zoneToUpdate, existingZone)
+      updateZoneChange <- ZoneChangeGenerator.forUpdate(zoneToUpdate, existingZone, auth).toResult
```
Comment (britneywright) [inline modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneService.scala]: how about `zoneWithUpdates` instead of `zoneToUpdate`

File: modules/api/functional_test/live_tests/zones/update_zone_test.py [hunk 14]
```diff
@@ -226,7 +227,8 @@ def test_update_zone_returns_404_if_zone_not_found(shared_zone_test_context):
             'keyName': VinylDNSTestContext.dns_key_name,
             'key': VinylDNSTestContext.dns_key,
             'primaryServer': VinylDNSTestContext.dns_ip
-        }
+        },
+        'adminGroupId': shared_zone_test_context.ok_group['id']
```
Comment (nimaeskandary) [inline modules/api/functional_test/live_tests/zones/update_zone_test.py]: Is this necessary?

File: modules/core/src/main/scala/vinyldns/core/domain/zone/Zone.scala [hunk 15]
```diff
@@ -88,6 +88,21 @@ object Zone {
...
       isTest = isTest)
   }
+
+  def apply(updateZoneInput: UpdateZoneInput, currentZone: Zone): Zone = {
+    import updateZoneInput._
+
+    currentZone.copy(
+      name = name,
+      email = email,
+      id = id,
```
Comment (nimaeskandary) [inline modules/core/src/main/scala/vinyldns/core/domain/zone/Zone.scala]: does this match up with what the json serializer accepts? seems the ability to update id was taken out

Comment (nimaeskandary) [reply modules/core/src/main/scala/vinyldns/core/domain/zone/Zone.scala]: nevermind about id, see that it was just moved in the code a bit

File: modules/api/src/main/scala/vinyldns/api/route/DnsJsonProtocol.scala [hunk 16]
```diff
@@ -105,26 +91,21 @@ trait DnsJsonProtocol extends JsonValidation {
       ).mapN(CreateZoneInput.apply)
   }
 
-  case object ZoneSerializer extends ValidationSerializer[Zone] {
-    override def fromJson(js: JValue): ValidatedNel[String, Zone] =
+  case object UpdateZoneInputSerializer extends ValidationSerializer[UpdateZoneInput] {
+    override def fromJson(js: JValue): ValidatedNel[String, UpdateZoneInput] =
       (
+        (js \ "id").required[String]("Missing Zone.id"),
```
Comment (nimaeskandary) [inline modules/api/src/main/scala/vinyldns/api/route/DnsJsonProtocol.scala]: will we reject updates that have all the other info or just drop those attributes?

File: modules/api/functional_test/live_tests/zones/update_zone_test.py (hunk 6 above)
Comment (mitruly) [reply modules/api/functional_test/live_tests/zones/update_zone_test.py]: Great suggestion - just tested create `vinyldns.` with shared flag set locally, and then updated to `false`.

File: modules/api/functional_test/live_tests/zones/update_zone_test.py (hunk 14 above)
Comment (mitruly) [reply modules/api/functional_test/live_tests/zones/update_zone_test.py]: Probably not. I'll revert.

File: modules/core/src/main/scala/vinyldns/core/domain/zone/Zone.scala (hunk 15 above)
Comment (mitruly) [reply modules/core/src/main/scala/vinyldns/core/domain/zone/Zone.scala]: Yeah, it was just moved around a bit. I don't believe that `id` could ever be updated since that is how we identify the zones (though `name` technically can be).

File: modules/api/src/main/scala/vinyldns/api/route/DnsJsonProtocol.scala (hunk 16 above)
Comment (mitruly) [reply modules/api/src/main/scala/vinyldns/api/route/DnsJsonProtocol.scala]: Any excess attributes in the deserialization from JSON just get dropped. Before we would accept them despite them being fields that we _shouldn't_ be able to change, so we are correcting that by limiting what we actually use from an update standpoint.

File: modules/api/functional_test/live_tests/zones/update_zone_test.py (hunk 14 above)
Comment (mitruly) [reply modules/api/functional_test/live_tests/zones/update_zone_test.py]: No, I realized that this is now required since `UpdateZoneInput` requires `adminGroupId` (there's conversation in #348 that refers to the `ZoneInput` model defined in #345).

Comment (nimaeskandary) [reply modules/api/functional_test/live_tests/zones/update_zone_test.py]: wow just realized i was reading this as owner group id, my bad foggy mind from shared zone work

File: modules/api/src/main/scala/vinyldns/api/repository/TestDataLoader.scala [hunk 17]
```diff
@@ -37,7 +37,8 @@ object TestDataLoader {
     firstName = Some("Test"),
     lastName = Some("User"),
     email = Some("test@test.com"),
-    isTest = true
+    isTest = true,
+    isSuper = true
```
Comment (rebstar6) [inline modules/api/src/main/scala/vinyldns/api/repository/TestDataLoader.scala]: nope nope cant load a super

Comment (mitruly) [reply modules/api/src/main/scala/vinyldns/api/repository/TestDataLoader.scala]: Good catch - thought this got reverted after testing locally.

File: modules/core/src/test/scala/vinyldns/core/TestZoneData.scala [hunk 18]
```diff
@@ -18,38 +18,67 @@ package vinyldns.core
...
     ZoneConnection("vinyldns.", "vinyldns.", "nzisn+4G2ldMn0q1CV3vsg==", "10.1.1.1"))
 
+  /* ZONES */
+  val okZone: Zone = Zone(
+    "ok.zone.recordsets.",
+    "test@test.com",
+    adminGroupId = okGroup.id,
+    connection = testConnection)
+  val abcZone: Zone = Zone("abc.zone.recordsets.", "test@test.com", adminGroupId = abcGroup.id)
+  val xyzZone: Zone = Zone("xyz.", "abc@xyz.com", adminGroupId = xyzGroup.id)
```
Comment (britneywright) [inline modules/core/src/test/scala/vinyldns/core/TestZoneData.scala]: Was this data (and related additions to this file) removed from the VinylDNSTestData? We should make sure so we can tell when the other file can be removed completely.

File: modules/docs/src/main/tut/api/zone-model.md [hunk 19]
```diff
@@ -30,9 +30,16 @@ account       | string      | **DEPRECATED** The account that created the zone |
 email         | string      | The distribution email for the zone |
 connection    | ZoneConnection | The connection used to issue DDNS updates to the backend zone.  If not provided, default keys will be used.  See the [Zone Connection Attributes](#zone-conn-attr) for more information |
 transferConnection | ZoneConnection | The connection that is used to sync the zone with the DNS backend.  This can be different than the update connection.  If not provided, default keys will be used |
+<<<<<<< HEAD
```
Comment (britneywright) [inline modules/docs/src/main/tut/api/zone-model.md]: remove

File: modules/docs/src/main/tut/api/zone-model.md [hunk 20]
```diff
@@ -30,9 +30,16 @@ account       | string      | **DEPRECATED** The account that created the zone |
...
+shared        | boolean     | An indicator that the zone is shared with anyone. Can only be set and/or modified by super users. Default value is `false`. |
+acl           | ZoneACL     | The access control rules governing the zone.  See the [Zone ACL Rule Attributes](#zone-acl-rule-attr) for more information
+id            | string      | The unique identifier for this zone
+isTest        | boolean     | An indicator for whether the zone is intended for testing purposes. |
+=======
 shared        | boolean     | **DEPRECATED** An indicator that the zone is shared with anyone.  No longer used as ACLs are now used to manage zones. |
 acl           | ZoneACL     | The access control rules governing the zone.  See the [Zone ACL Rule Attributes](#zone-acl-rule-attr) for more information
 id            | string      | The unique identifier for this zone
+>>>>>>> parent of e6201006... Update docs.
```
Comment (britneywright) [inline modules/docs/src/main/tut/api/zone-model.md]: remove

File: modules/docs/src/main/tut/api/zone-model.md [hunk 21]
```diff
@@ -30,9 +30,16 @@ account       | string      | **DEPRECATED** The account that created the zone |
 email         | string      | The distribution email for the zone |
 connection    | ZoneConnection | The connection used to issue DDNS updates to the backend zone.  If not provided, default keys will be used.  See the [Zone Connection Attributes](#zone-conn-attr) for more information |
 transferConnection | ZoneConnection | The connection that is used to sync the zone with the DNS backend.  This can be different than the update connection.  If not provided, default keys will be used |
+<<<<<<< HEAD
+shared        | boolean     | An indicator that the zone is shared with anyone. Can only be set and/or modified by super users. Default value is `false`. |
+acl           | ZoneACL     | The access control rules governing the zone.  See the [Zone ACL Rule Attributes](#zone-acl-rule-attr) for more information
+id            | string      | The unique identifier for this zone
+isTest        | boolean     | An indicator for whether the zone is intended for testing purposes. |
+=======
```
Comment (britneywright) [inline modules/docs/src/main/tut/api/zone-model.md]: remove

File: modules/core/src/main/scala/vinyldns/core/domain/zone/Zone.scala (hunk 15 above)
Comment (rebstar6) [reply modules/core/src/main/scala/vinyldns/core/domain/zone/Zone.scala]: yeah interesting, ID should stay static, so no need to copy here

Comment (rebstar6) [reply modules/core/src/main/scala/vinyldns/core/domain/zone/Zone.scala]: name really shouldnt change either. that doesnt make much sense...

for now keep taht in the copy though, we should stop a name update in the service (or maybe not even accept it in the update). followup though, not now
